src/c/inteldeflater/IntelDeflater.c

```diff
@@ -116,20 +201,65 @@ Java_htsjdk_samtools_util_zip_IntelDeflater_setDictionary(JNIEnv *env, jclass cl
 }
 
 JNIEXPORT jint JNICALL
-Java_htsjdk_samtools_util_zip_IntelDeflater_deflateBytes(JNIEnv *env, jobject this, jlong addr,
+Java_net_sf_samtools_util_zip_IntelDeflater_deflateBytes(JNIEnv *env, jobject this, jlong addr,
                                          jarray b, jint off, jint len, jint flush)
 {
-    z_stream *strm = jlong_to_ptr(addr);
+  jarray this_buf = (*env)->GetObjectField(env, this, bufID);
+  jint this_off = (*env)->GetIntField(env, this, offID);
+  jint this_len = (*env)->GetIntField(env, this, lenID);
+  jbyte *in_buf;
+  jbyte *out_buf;
+  LZ_Stream2 *strm;
+  if (useIGZIP)
+    strm = jlong_to_ptr(addr);
+  //igzip only supports one compression level so setParamsID should not be set when using igzip 
+  //igzip does not support flush
+  if (useIGZIP && (((*env)->GetBooleanField(env, this, setParamsID) && strm->total_in != 0)  || flush == 1)) {
+    JNU_ThrowInternalError(env, "igzip doesn't support this");
+  } else if (useIGZIP) {
+    in_buf = (*env)->GetPrimitiveArrayCritical(env, this_buf, 0);
+    if (in_buf == NULL) {
+      // Throw OOME only when length is not zero
+      if (this_len != 0)
+	JNU_ThrowOutOfMemoryError(env, 0);
+      return 0;
+    }
+    out_buf = (*env)->GetPrimitiveArrayCritical(env, b, 0);
+    if (out_buf == NULL) {
+      (*env)->ReleasePrimitiveArrayCritical(env, this_buf, in_buf, 0);
+      if (len != 0)
+	JNU_ThrowOutOfMemoryError(env, 0);
+      return 0;
+    }
+    strm->next_in = (Bytef *) (in_buf + this_off);
+    strm->next_out = (Bytef *) (out_buf + off);
+    strm->avail_in = this_len;
+    strm->avail_out = len;
+    assert(strm->avail_in != 0);
+    assert(strm->avail_out != 0);
+    jboolean finish = (*env)->GetBooleanField(env, this, finishID);
+    if (finish) 
+      strm->end_of_stream = 1;
+    else
+      strm->end_of_stream = 0;
+
+    fast_lz(strm);
```

Comment:
just curious, if `lz_stream.avail_in > 0` here does that mean that there was not enough room in the block for the compressed bytes?  If so, perhaps it should default to using `zlib`?  See a similar thought here: https://github.com/nh13/pbgzip/blob/master/consumer.c#L297